lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -252,6 +254,126 @@ void FIRRTLTypesLowering::visitDecl(InstanceOp op) {
 }
 
+void FIRRTLTypesLowering::visitDecl(MemOp op) {
+
+  auto type = op.getDataTypeOrNull();
+  // Bail out if the memory is empty
+  if (!type)
+    return;
+
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenBundleTypes(type, "", false, fieldTypes);
+
+  // Mutable store of the types of the ports of a new memory
+  SmallVector<Type, 4> resultPortTypes;
+
+  DenseMap<StringRef, Value> newWires;
+  for (auto field : fieldTypes) {
+    // Create the new type for the memory. Copy every field, but
+    // update the "data" field to be the ground type and the "mask"
+    // field to be a uint<1> type.
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+      SmallVector<BundleType::BundleElement, 5> resultFields;
+
+      bool flipped = op.getResult(i).getType().isa<FlipType>();
+      BundleType underlying = op.getResult(i)
+                                  .getType()
+                                  .cast<FIRRTLType>()
+                                  .getPassiveType()
+                                  .cast<BundleType>();
+
+      for (auto elt : underlying.getElements()) {
+        auto newElt =
+            StringSwitch<BundleType::BundleElement>(elt.name)
+                .Cases("clk", "en", "addr", "wmode", elt)
+                .Cases("mask", "wmask",
+                       {elt.name, UIntType::get(builder->getContext(), 1)})
+                .Default(/* "data", "rdata", "wdata" */
+                         {elt.name, field.type});
+        resultFields.push_back(newElt);
+      }
+
+      if (flipped) {
+        resultPortTypes.push_back(FlipType::get(
+            BundleType::get(resultFields, builder->getContext())));
+        continue;
+      }
+
+      resultPortTypes.push_back(
+          BundleType::get(resultFields, builder->getContext()));
+    }
+
+    // Construct the new memory for this flattened field
+    auto newName = op.name().getValue().str() + field.suffix;
+    auto newMem = builder->create<MemOp>(
+        resultPortTypes, op.readLatency(), op.writeLatency(), op.depth(),
+        op.ruw(), op.portNames(), builder->getStringAttr(newName));
+    resultPortTypes.clear();
+
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+
+      bool flipped = op.getResult(i).getType().isa<FlipType>();
+      BundleType underlying = op.getResult(i)
+                                  .getType()
+                                  .cast<FIRRTLType>()
+                                  .getPassiveType()
+                                  .cast<BundleType>();
+
+      for (auto elt : underlying.getElements()) {
+        auto index = StringSwitch<int>(elt.name)
+                         .Cases("clk", "en", "addr", "wmode", 0)
+                         .Cases("mask", "wmask", 1)
+                         .Default(/* "data", "rdata", "wdata", */ 2);
+        switch (index) {
+        case 0: {
+          Type theType = elt.type;
+          if (flipped)
+            theType = FlipType::get(elt.type);
```

Comment:
I was using the (slower) `SubfieldOp::getResultType(op.getResult(i), field.name, op.getLoc())` to get this done.  I'm not totally sure what I'm doing.  Maybe we should factor this out into a helper that can get the field type by index (it could work for vectors?)?